### PR TITLE
[Facilities] Add facility filters for each location

### PR DESCRIPTION
**Pull request by linyc0817**
# Description
When click into a cluster, facility markers will show nothing in default. 
After clicking the filter buttons above, markers of the corresponding category will be displayed.

# Changes
## Filter Fabs
1. Filter categories depends on the cluster json file
2. Multiple filter buttons can be selected at the same time 

## Facility markers
1. Markers displayed only when corresponding category button been activated

截圖 2023-11-08 下午10 47 43

截圖 2023-11-08 下午10 48 07

# Notes
Both facility markers and facility filters will show only when its cluster is open

# Checklist

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] Any changes to strings have been published to our translation tool
- [x] I conducted basic QA to assure all features are working
- [x] I requested code review from other team members

# Resolved Issues
#12

### CODE REVIEW

File: src/models/facility/index.ts [hunk 1]
```diff
@@ -2,7 +2,6 @@ interface Facility {
   id: string;
   data: FacilityData;
 }
```
Comment (y-c-wang) [inline src/models/facility/index.ts:4]:
```suggestion
}

```

File: src/constants/facility.ts [hunk 2]
```diff
@@ -40,4 +40,15 @@ function getLocations(clusterName: string): Facilities {
...
-export { getLocations };
+function getLocationCategories(clusterName: string) {
+  const allFacilities: Facilities = getLocations(clusterName);
+  const categories = new Set<string>();
+
+  for (const facility of Object.values(allFacilities)) {
+    categories.add(facility.target.name);
+  }
+
+  return Array.from(categories);
```
Comment (y-c-wang) [inline src/constants/facility.ts:51]:
```suggestion
  const facilities: Facilities = getLocations(clusterName);
  const categories = Object.values(facilities).map((facility) => facility.target.name);

  return Array.from(categories);
```

File: src/components/Map/Fabs/FacilityFilterFab/index.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import { useGetClusterQuery } from "../../../../api/cluster";
+import FacilityMarker from "../../Markers/FacilityMarkers";
+
+const FacilityFilterFabs: React.FC = () => {
+  const [searchParams] = useSearchParams();
+
+  const clusterId = getParamsFromDrawer("cluster", searchParams).clusterId;
+  const { data: cluster } = useGetClusterQuery(clusterId);
+
+  const allCategories = cluster ? getLocationCategories(cluster.data.name) : [];
```
Comment (y-c-wang) [inline src/components/Map/Fabs/FacilityFilterFab/index.tsx:15]:
```suggestion
  const categories = cluster ? getLocationCategories(cluster.data.name) : [];
```

File: src/components/Map/Fabs/FacilityFilterFab/index.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import React, { useEffect } from "react";
+import { Button } from "@nextui-org/react";
+import { getLocationCategories } from "../../../../constants/facility";
+import { useSearchParams } from "react-router-dom";
+import { getParamsFromDrawer } from "../../../../utils/routes/params";
+import { useGetClusterQuery } from "../../../../api/cluster";
+import FacilityMarker from "../../Markers/FacilityMarkers";
```
Comment (y-c-wang) [inline src/components/Map/Fabs/FacilityFilterFab/index.tsx:7]:
```suggestion
import FacilityMarkers from "../../Markers/FacilityMarkers";
```

Seems to be a typo

File: src/components/Map/Markers/index.tsx [hunk 5]
```diff
@@ -1,15 +1,13 @@
...
 import ClusterMarkers from "./ClusterMarkers";
-import FacilityMarkers from "./FacilityMarkers";
 import PoiMarkers from "./PoiMarkers";
 import UserMarker from "./UserMarker";
 
 const Markers: React.FC = () => {
   return (
     <>
       <ClusterMarkers />
-      <FacilityMarkers />
```
Comment (y-c-wang) [inline src/components/Map/Markers/index.tsx:12]: Please do not move any markers out of this file

**Review by y-c-wang — CHANGES_REQUESTED** [review]

File: src/components/Map/Markers/index.tsx (hunk 5 above)
Comment (linyc0817) [reply src/components/Map/Markers/index.tsx:12]: Even though i already contain it in the FacilityFilterFab? 
Won't duplicate cause problems?

File: src/components/Map/Fabs/FacilityFilterFab/index.tsx (hunk 3 above)
Comment (linyc0817) [reply src/components/Map/Fabs/FacilityFilterFab/index.tsx:15]: done

File: src/models/facility/index.ts (hunk 1 above)
Comment (linyc0817) [reply src/models/facility/index.ts:4]: done

File: src/components/Map/Fabs/FacilityFilterFab/index.tsx (hunk 4 above)
Comment (linyc0817) [reply src/components/Map/Fabs/FacilityFilterFab/index.tsx:7]: FacilityMarkers components remove from FacilityFilterFab !

File: src/components/Map/Markers/index.tsx (hunk 5 above)
Comment (linyc0817) [reply src/components/Map/Markers/index.tsx:12]: Using redux to store the selectedCategories

File: src/constants/facility.ts (hunk 2 above)
Comment (y-c-wang) [reply src/constants/facility.ts]: Thanks for suggestion. Let's consider using `reduce` here :)

**Review by y-c-wang — APPROVED** [review]

**Pull request merged**